scripts/ci/render_bench.py

```diff
@@ -0,0 +1,380 @@
+#!/usr/bin/env python3
+
+"""
+Render benchmark graphs from data in GCS.
+
+To use this script, you must be authenticated with GCS,
+see https://cloud.google.com/docs/authentication/client-libraries for more information.
+
+Install dependencies:
+    google-cloud-storage==2.9.0
+
+Use the script:
+    python3 scripts/ci/render_bench.py --help
+
+    python3 scripts/ci/render_bench.py \
+      crates \
+      --num-days 30 \
+      --output ./benchmarks
+
+    python3 scripts/ci/render_bench.py \
+      sizes \
+      --num-days $((30*6)) \
+      --output gs://rerun-builds/graphs
+"""
+from __future__ import annotations
+
+import argparse
+import json
+import os
+import re
+import textwrap
+from dataclasses import dataclass
+from datetime import datetime, timedelta
+from enum import Enum
+from pathlib import Path
+from subprocess import run
+from typing import Any, Callable, Dict, Generator, List
+
+from google.cloud import storage
+
+SCRIPT_PATH = os.path.dirname(os.path.relpath(__file__))
+DATE_FORMAT = "%Y-%m-%d"
+ESCAPED_DATE_FORMAT = DATE_FORMAT.replace("%", "%%")
+
+
+def non_empty_lines(s: str) -> Generator[str, None]:
+    for line in s.splitlines():
+        if len(line.strip()) == 0:
+            continue
+        yield line
+
+
+def git_log(date: datetime) -> list[str] | None:
+    args = ["git", "log"]
+    args += [f'--after="{date.year}-{date.month}-{date.day} 00:00:00"']
+    args += [f'--before="{date.year}-{date.month}-{date.day} 23:59:59"']
+    args += ["--format=%H"]
+    commits = run(args, check=True, capture_output=True, text=True).stdout.strip()
+    if len(commits) == 0:
+        return None
+    else:
+        return commits.strip().splitlines()
+
+
+CommitsByDate = Dict[datetime, List[str]]
+
+
+def get_commits(end_date: datetime, num_days: int) -> CommitsByDate:
+    """Yields the list of commits for each day between `end_date - num_days` and `end_date`."""
+
+    start = (end_date - timedelta(days=num_days)).strftime("%Y-%m-%d")
+    end = (end_date).strftime("%Y-%m-%d")
+    print(f"Fetching commits between {start} and {end}")
+
+    commits_by_date = {}
+    previous_log = None
+
+    for offset in reversed(range(num_days)):
+        date = end_date - timedelta(days=offset)
+        log = git_log(date) or previous_log
+        if log is None:
+            continue
+        commits_by_date[date] = log
+        previous_log = log
+
+    return commits_by_date
```

Comment:
I get it makes for a nicer graph, but I actually doubt the practicality of doing a sample per day. Most of the time I'm more interested in commit to commit changes. But maybe the resulting varying density and amount of data is too much to ask of the plotting lib here?

If this is too problematic we could also have a mode for the "n last commits". The way you set this script up it seems pretty easy to have both - for overview the days is useful, for research into recent results per commit is likely better.

Reply:
No, I think chartjs can handle up to a few thousand points per graph here. It'll actually be harder to look for specific commits if the points are too small.

I'll put _every_ available commit in the last 30 days on the graphs, and add a "zoom in" button which will put it in a full-screen modal.

Reply:
that sounds great :)